source/matrix_free/task_info.cc

```diff
@@ -866,93 +813,222 @@ namespace internal
+      // bins, those which we know can be grouped together by the given parent
+      // relation and those which cannot
+      std::vector<unsigned int> temporary_numbering;
+      temporary_numbering.reserve(n_active_cells +
+                                  (n_lanes - 1) * n_categories);
+      const unsigned int n_cells_per_parent =
+        std::count(parent_relation.begin(), parent_relation.end(), 0);
+      std::vector<unsigned int> category_size;
+      for (unsigned int j = 0; j < n_categories; ++j)
+        {
+          std::vector<std::pair<unsigned int, unsigned int>> grouped_cells;
+          std::vector<unsigned int>                          other_cells;
+          for (const unsigned int cell : renumbering_category[j])
+            if (parent_relation.empty() ||
+                parent_relation[cell] == numbers::invalid_unsigned_int)
+              other_cells.push_back(cell);
+            else
+              grouped_cells.emplace_back(parent_relation[cell], cell);
+
+          // Compute the number of cells per group
+          std::sort(grouped_cells.begin(), grouped_cells.end());
+          std::vector<unsigned int> n_cells_per_group;
+          unsigned int              length = 0;
+          for (unsigned int i = 0; i < grouped_cells.size(); ++i, ++length)
+            if (i > 0 && grouped_cells[i].first != grouped_cells[i - 1].first)
+              {
+                n_cells_per_group.push_back(length);
+                length = 0;
+              }
+          if (length > 0)
+            n_cells_per_group.push_back(length);
+
+          // Move groups that do not have the complete size (due to
+          // categories) to the 'other_cells'. The cells with correct group
+          // size are immediately appended to the temporary cell numbering
+          auto group_it = grouped_cells.begin();
+          for (unsigned int length : n_cells_per_group)
+            if (length < n_cells_per_parent)
+              for (unsigned int j = 0; j < length; ++j)
+                other_cells.push_back((group_it++)->second);
+            else
+              {
+                // we should not have more cells in a group than in the first
+                // check we did above
+                AssertDimension(length, n_cells_per_parent);
+                for (unsigned int j = 0; j < length; ++j)
+                  temporary_numbering.push_back((group_it++)->second);
+              }
+
+          // Sort the remaining cells and append them as well
+          std::sort(other_cells.begin(), other_cells.end());
+          temporary_numbering.insert(temporary_numbering.end(),
+                                     other_cells.begin(),
+                                     other_cells.end());
+
+          while (temporary_numbering.size() % n_lanes != 0)
+            temporary_numbering.push_back(numbers::invalid_unsigned_int);
+
+          category_size.push_back(temporary_numbering.size());
```

Comment:
If I read this code correctly, this code enumerates cells differently than in the version before. But this is fixed with the sort below?

Reply:
Yes, we enumerate cells differently now (which is why the tests sensitive to those have changed).
Here we first place the "structured" part and the "unstructured" one. The sort below allows to set priorities on how to arrange cell batches when combining them. Right now we have two priorities:
- In the hp case, cells with the same degree are placed together (because the tests depend on that for now; it could be relaxed, though)
- We keep the cell batches sorted by the index of the last cell. This makes sure that we make a good decision how the two queues are combined.
Do you think we should add something like that as an overall comment?

Reply:
> Do you think we should add something like that as an overall comment?

Comments are always good! It took me a couple of seconds to understand why `std::array<..., 3>`. I like this approach it (and the padding with `numbers::invalid_unsigned_int`) should more or less have the same effect as the stat machine from the weekend.

> Yes, we enumerate cells differently now

My question was referring to the version on the weekend. But anyhow, I am fine with the changes.